contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -481,13 +510,62 @@ abstract contract BaseBalancerStrategy is InitializableAbstractStrategy {
 
     function _approveBase() internal virtual {
         IERC20 pToken = IERC20(platformAddress);
+
         // Balancer vault for BPT token (required for removing liquidity)
-        pToken.safeApprove(address(balancerVault), type(uint256).max);
+        // slither-disable-next-line unused-return
+        pToken.approve(address(balancerVault), 0);
+        // slither-disable-next-line unused-return
+        pToken.approve(address(balancerVault), type(uint256).max);
+    }
+
+    function cacheRateProviders() public {
+        /* Rate providers haven't been cached yet. It is OK to not
+         * expire the rate providers cache, since those are not
+         * expected to ever change.
+         */
+        IRateProvider[] memory providers = IBalancerPool(platformAddress)
+            .getRateProviders();
+        poolRateProvidersCache = providers;
     }
 
+    /**
+     * @notice Returns the rate supplied by the Balancer configured rate
+     * provider. Rate is used to normalize the token to common underlying
+     * pool denominator. (ETH for ETH Liquid staking derivatives).
+     *
+     * @param _asset Address of the Balancer pool asset
+     * @return rate of the corresponding asset
+     */
     function _getRateProviderRate(address _asset)
```

Comment:
Does this method "work" and return a possibly spurious value for assets which are not supported? 

`poolAssetIndex` is mapping, so a bad _asset value should return a zero. This zero would then be used as a valid first index into the rate provider, so everything that's not valid, non-first pool assets would return the rate of the first asset?

We would need to make sure then that only valid addresses are fed into this method.

----

This current method does three storage slot reads.

1. poolAssetIndex mapping lookup
2. poolRateProvidersCache length bounds check
3. poolRateProvidersCache array lookup

But if this method is the only method that uses the `poolRateProvidersCache`, then we could trim this down to one read by making `poolRateProvidersCache` an address => address mapping from the _asset directly to the rate provider.

Also, if we used a special value (say `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE`) for a fixed exchange rate, then we could revert on unsupported addresses, instead of silently continuing with them.

Reply:
awesome thank you, this is a great find and a good proposal

Reply:
So this has moved into a separate PR because it has more profound changes that enable testing of the internal functions: https://github.com/OriginProtocol/origin-dollar/pull/1877